### PR TITLE
Add To Cart with Options: Register store for the product types

### LINKED ISSUE #53365: Add To Cart with Options: Implement Data propagation for the product type
When composing the Add To Cart section at the template level, the app should show a product type selector that will help the user see how the block behaves or looks depending on the product type.
Let's store that globally.

### PR DESCRIPTION
**Pull request by retrofox**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR introduces the `woocommerce/template-state` store to handle the global product type.
The store could scale to handle other global variables, such as product quantity, and, according to it, change the block representation.

Closes #53365

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

You need to use the Redux Dev Tool ([chrome](https://chromewebstore.google.com/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd?hl=pt-PT), [firefox](https://chromewebstore.google.com/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd?hl=pt-PT))

1. Install and activate the WooCommerce Beta Tester plugin
2. Go to WooCommerce Admin Test Helper -> Features and enable `blockified-add-to-cart`
3. Create a new product.
4. Go to Pages > Add New Page.
5. Open the Redux dev tool
6. Filter the store by the `woocommerce/product-template` name

Screenshot 2024-12-03 at 12 25 53 PM

7. Open the dev console
8. Load the product types dispatching the proper action:

```es6
wp.data.dispatch( 'woocommerce/template-state' ).setProductTypes( [
    {
        slug: 'simple',
        label: 'Simple Product',
    },
    {
        slug: 'variable',
        label: 'Variable Product',
    },
    {
        slug: 'grouped',
        label: 'Grouped Product',
    },
] );
```

9. Confirm the store populates

Screenshot 2024-12-03 at 1 07 36 PM

11. Select the current type dispatching the following action:

```es6
wp.data.dispatch( 'woocommerce/template-state' ).switchProductType( 'grouped' );
```

12. Confirm it populate the state:

Screenshot 2024-12-03 at 12 29 33 PM

13. Get the current product type by using the following selector:

```es6
wp.data.select( 'woocommerce/template-state' ).getCurrentProductType()
```

image

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/types.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+export type ProductTypeSlug = 'single-product' | string;
```
Comment (Aljullu) [inline plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/types.ts]: What does `single-product` mean here? Should it be used as a generic product type when we don't know it?

File: plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/utils/get-product-types/index.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
+/**
+ * External dependencies
+ */
+import { getSetting } from '@woocommerce/settings';
+
+export type ProductTypeProps = {
```
Comment (Aljullu) [inline plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/utils/get-product-types/index.ts:6]: We now have two `ProductTypeProps` exported from two different files. Could we consolidate this into one?

**Review by Aljullu — APPROVED** [review]
LGTM, thanks @retrofox! I left two minor comments but nothing blocking.

~I guess after this PR and #53354 are merged, we will still need a follow-up PR to connect the product type selector with the store, is that correct?~ Nevermind, I just saw https://github.com/woocommerce/woocommerce/issues/53389. :smile:

File: plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/types.ts (hunk 1 above)
Comment (retrofox) [reply plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/types.ts]: it doesn't make sense. Updating...

File: plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/utils/get-product-types/index.ts (hunk 2 above)
Comment (retrofox) [reply plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/utils/get-product-types/index.ts:6]: let me address this mess

**Pull request merged**